File: src/data/long_distance_social_distancing_ii.java
import java.io.*;
import java.util.*;

public class Solution {
    public static long mod = 1000000007;

    public static long[][] matmul(long[][] a, long[][] b) {
        long[][] c = new long[5][5];
        for (int i = 0; i < 5; i++) {
            for (int j = 0; j < 5; j++) {
                c[i][j] = 0;
                for (int k = 0; k < 5; k++) {
                    c[i][j] += (a[i][k] * b[k][j]) % mod;
                    c[i][j] %= mod;
                }
            }
        }
        return c;
    }

    public static long[] matmul(long[][] a, long[] b) {
        long[] c = new long[5];
        for (int i = 0; i < 5; i++) {
            for (int j = 0; j < 5; j++) {
                c[i] += (a[i][j] * b[i]) % mod;
                c[i] %= mod;
            }
        }
        return c;
    }

    public static long[][] pow(long[][] mat, long p) {
        if (p == 0)
            return new long[][]{{1, 0, 0, 0, 0}, {0, 1, 0, 0, 0}, {0, 0, 1, 0, 0}, {0, 0, 0, 1, 0}, {0, 0, 0, 0, 1}};
        long[][] temp = pow(mat, p / 2);
        if (p % 2 == 0) {
            return matmul(temp, temp);
        }
        return matmul(matmul(temp, temp), mat);
    }

    public static long solve(long t) {
        if (t == 0) return 1;
        long[][] init = new long[][]{
                {1, 1, 1, 1, 1},
                {1, 0, 1, 1, 0},
                {1, 1, 0, 1, 1},
                {1, 1, 1, 0, 0},
                {1, 0, 1, 0, 0}
        };
        long[][] result = pow(init, t - 1);
        long[] lastRow = matmul(result, new long[]{1, 1, 1, 1, 1});
        long s = 0;
        for (int i = 0; i < 5; i++) {
            s += lastRow[i];
            s %= mod;
        }
        return (s * s) % mod;
    }

    public static void main(String[] args) throws Throwable {
        /**
         * This is the most direct way of implementing the algorithm described. You can optimize by caching some of the exponentiation 
         * between each test case and by doing matrix multiplication in place.
         */
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        long t = Integer.parseInt(reader.readLine());
        for (int i = 0; i < t; i++) System.out.println(solve(Long.parseLong(reader.readLine())));
        reader.close();
    }
}

File: src/data/long_distance_social_distancing_ii.cpp
#include <string>
#include <iterator>
#include <iostream>
#include <algorithm>
#include <array>

using namespace std;
#define LL long long
#define UL unsigned long
#define matrix array<array<LL, 5>, 5>
#define vector array<LL, 5>

LL mod = 1000000007;

matrix matmul(matrix a, matrix b)
{
    matrix c;
    for (unsigned long i = 0; i < a.size(); i++)
    {
        for (unsigned long j = 0; j < b.size(); j++)
        {
            c[i][j] = 0;
            for (unsigned long k = 0; k < 5; k++)
            {
                c[i][j] += (a[i][k] * b[k][j]) % mod;
                c[i][j] %= mod;
            }
        }
    }
    return c;
}

vector matmul(matrix a, vector b)
{
    vector c;
    for (int i = 0; i < 5; i++)
    {
        c[i] = 0;
        for (int j = 0; j < 5; j++)
        {
            c[i] += (a[i][j] * b[i]) % mod;
            c[i] %= mod;
        }
    }
    return c;
}

matrix pow(matrix mat, long p)
{
    if (p == 0)
        return matrix{{{1, 0, 0, 0, 0}, {0, 1, 0, 0, 0}, {0, 0, 1, 0, 0}, {0, 0, 0, 1, 0}, {0, 0, 0, 0, 1}}};
    matrix temp = pow(mat, p / 2);
    if (p % 2 == 0)
        return matmul(temp, temp);
    return matmul(matmul(temp, temp), mat);
}

LL solve(LL t)
{
    if (t == 0)
        return 1;
    matrix init = matrix{{{1, 1, 1, 1, 1},
                          {1, 0, 1, 1, 0},
                          {1, 1, 0, 1, 1},
                          {1, 1, 1, 0, 0},
                          {1, 0, 1, 0, 0}}};
    matrix result = pow(init, t - 1);
    vector lastRow = matmul(result, vector{1, 1, 1, 1, 1});
    LL s = 0;
    for (int i = 0; i < 5; i++)
    {
        s += lastRow[i];
        s %= mod;
    }
    return (s * s) % mod;
}

int main()
{
    int t;
    LL c;
    cin >> t;
    for (int i = 0; i < t; i++)
    {
        cin >> c;
        cout << solve(c) << endl;
    }
}